### PR TITLE
Added support for devices

### LINKED ISSUE #6: Add support for Devices in Lookup
- Add support for Devices in Lookup

### PR DESCRIPTION
**Pull request by codeMinter**
Fixes #6

### CODE REVIEW

**Comment by callmekatootie** [comment]
Unit tests fail:

image

**Comment by callmekatootie** [comment]
E2E tests fail too:

image

**Comment by codeMinter** [comment]
hey @callmekatootie seems you are testing wrong PR. This is not ready hence marked not to merge.

**Comment by codeMinter** [comment]
hey @callmekatootie this is ready for merge. Kindly validate once your side too and let's merge and backend ES processor asap.

File: test/unit/helper.test.js [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+      } catch (err) {
+
+      }
+    })
+
+    it('Delete Table success', async () => {
+      await service.deleteTable(testTableName)
+    })
+  })
+})
```
[the remark this answers is not in the document]
Comment (callmekatootie) [reply test/unit/helper.test.js]: Also - I DONT think you need to provide the helper file unit tests

File: README.md [hunk 2]
```diff
@@ -58,18 +55,17 @@ The following test parameters can be set in config file or in env variables:
 - M2M_READ_ACCESS_TOKEN: M2M read access token
 - M2M_UPDATE_ACCESS_TOKEN: M2M update access token
 - AMAZON.DYNAMODB_COUNTRY_TABLE: The country DynamoDB table used during unit/e2e tests (`test_` is appended to the table name automatically)
+- AMAZON.DYNAMODB_DEVICE_TABLE: The device DynamoDB table used during unit/e2e tests (`test_` is appended to the table name automatically)
 - AMAZON.DYNAMODB_EDUCATIONAL_INSTITUTION_TABLE: The educational institutions DynamoDB table used during unit/e2e tests (`test_` is appended to the table name automatically)
 - ES.COUNTRY_INDEX: The country elasticsearch index used during unit/e2e tests.
 - ES.COUNTRY_TYPE: The country elastic search document type used during unit/e2e tests.
+- ES.DEVICE_INDEX: The device elasticsearch index used during unit/e2e tests.
+- ES.DEVICE_TYPE: The device elastic search document type used during unit/e2e tests.
```
[the remark this answers is not in the document]
Comment (codeMinter) [reply README.md]: ok pushing in new commit.

File: README.md [hunk 3]
```diff
@@ -106,38 +98,39 @@ may be different for different ES versions.
...
+    > **Example:** `npm run load-data --lookup=devices --file=resources/devices.json`
+
+- To delete data from Lookup DynamoDB table `npm run clean-data --lookup=<lookup>` **WARNING: This deletes all the data in the lookup**
+    - `<lookup>` is the name of the lookup. For example: `devices`.
+    > **Example:** `npm run clean-data --lookup=devices`
+
+- **Supported Lookup:**
+    - countries
+    - educationalInstitutions
+    - devices
```
[the remark this answers is not in the document]
Comment (codeMinter) [reply README.md]: we will have json being prepared from xls, which will be updated as new devices are to be pushed. So load helps do it, but then thought what if we need to clean due to issue, so these both help there. Good to have options. `load-data` will load only unique values, as we evolve the list.

File: src/common/helper.js [hunk 4]
```diff
@@ -263,20 +288,48 @@ async function createESIndex (indexName) {
...
+      },
+      operatingSystemVersion: {
+        type: 'text',
+        fielddata: true
+      }
+
+    }
+  }
+
+  const ind = {
```
[the remark this answers is not in the document]
Comment (codeMinter) [reply src/common/helper.js]: oh yes, will make the change.

File: src/controllers/DeviceController.js [hunk 5]
```diff
@@ -0,0 +1,99 @@
+/**
+ * Controller for Device endpoints
+ */
+const HttpStatus = require('http-status-codes')
+const service = require('../services/DeviceService')
+const helper = require('../common/helper')
+
+/**
+ * List countries
```
[the remark this answers is not in the document]
Comment (codeMinter) [reply src/controllers/DeviceController.js]: done

File: README.md [hunk 6]
```diff
@@ -106,38 +98,39 @@ may be different for different ES versions.
...
 Note that running tests will clear all DynamoDB data in test tables and re-create Elasticsearch test indices.
 
 ### Running unit tests
-
 To run unit tests alone
-
-```bash
-npm run test
-```
+> npm run test
```
[the remark this answers is not in the document]
Comment (codeMinter) [reply README.md]: ok, was using online editor so trying to make these commands look at block, but no issues, putting them in same format as earlier.

File: test/unit/helper.test.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (codeMinter) [reply test/unit/helper.test.js]: this is weird, they both pass on myside
npm run test
> 209 passing (6s)

npm run e2e
> 245 passing (6s)

Can you give a try again?

[the remark this answers is not in the document]
Comment (codeMinter) [reply test/unit/helper.test.js]: ok try now, as the tests changed a bit due to common logic change for indexing.

**Comment by codeMinter** [comment]
@callmekatootie thanks - addressed review comments. For tests, did try both test and e2e and both work fine on myside.

**Comment by callmekatootie** [comment]
@codeMinter Tests still fail - I believe there's a race condition that's happening since sometimes 1 test fails and sometimes 2

image

**Comment by callmekatootie** [comment]
@codeMinter My advice is to drop the tests in the helper. They are not needed

**Comment by codeMinter** [comment]
> @codeMinter My advice is to drop the tests in the helper. They are not needed

@callmekatootie  Ok Done. Give a try now.

**Pull request merged**